daemon/sync.go

```diff
@@ -0,0 +1,484 @@
+package daemon
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"strings"
+	"time"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/cluster"
+	"github.com/weaveworks/flux/event"
+	"github.com/weaveworks/flux/git"
+	fluxmetrics "github.com/weaveworks/flux/metrics"
+	"github.com/weaveworks/flux/resource"
+	fluxsync "github.com/weaveworks/flux/sync"
+	"github.com/weaveworks/flux/update"
+)
+
+// Sync holds the data we are working with during a sync.
+type Sync struct {
+	logger      log.Logger
+	working     *git.Checkout
+	oldTagRev   string
+	newTagRev   string
+	initialSync bool
+	commits     []git.Commit
+
+	*Daemon
+}
+
+// NewSync prepares a sync.
+func (d *Daemon) NewSync(logger log.Logger) Sync {
+	s := Sync{logger: logger, Daemon: d}
+
+	return s
+}
+
+// Run starts the sync.
+func (s *Sync) Run(ctx context.Context) (retErr error) {
+	started := time.Now().UTC()
+	defer func() {
```

Comment:
I reckon do this around the method call, rather than in the method call, if possible.